pubtools/_content_gateway/push_cgw.py

```diff
@@ -1,2 +1,196 @@
+        version_id = self.pv_mapping.get((product_name, product_code, version_name)) if self.pv_mapping else None
+        if version_id:
+            _, version_id = version_id
+        return version_id
+
+    def __get_file_id(self, file_item):
+        product_name = file_item.get('metadata').get('productName')
+        product_code = file_item.get('metadata').get('productCode')
+        version_name = file_item.get('metadata').get('productVersionName')
+
+        product_id = self.__get_product_id(product_name, product_code)
+        if not self.pv_mapping:
+            self.__create_product_version_mapping(product_name, product_code)
+        version_id = self.__get_version_id(product_name, product_code, version_name)
+        all_files = self.cgw_client.get_files(product_id, version_id)
+        file_mapping = {(product_name, product_code, version_name, data.get('downloadURL')): (
+            product_id, version_id, data.get('id')) for data in all_files}
+        pvf_id = file_mapping.get(
+            (product_name, product_code, version_name,
+             file_item.get('metadata').get('downloadURL'))) if file_mapping else (None, None, None)
+        return pvf_id
+
+    def process_product(self, item):
+        try:
+            product_name = item.get('metadata').get('name')
+            product_code = item.get('metadata').get('productCode')
+            LOG.debug(f"Fetching for product_id of product name- {product_name} and product code- {product_code}")
+            product_id = self.__get_product_id(product_name, product_code)
+
+            if item.get('state') == 'present':
+                if not product_id:
+                    LOG.info("No previous entries found for the given product name and product code")
+                    LOG.info("Creating product entry for the given product metadata")
+                    product_id = self.cgw_client.create_product(item.get('metadata'))
+                    # A new product is created so updating product mapping
+                    cgw_data = self.cgw_client.get_products()
+                    self.product_mapping = {(data['name'], data['productCode']): data['id'] for data in cgw_data}
+                    LOG.debug(f"Created a new product with product_id:- {product_id}")
+                    return
+                LOG.info(f"Product record found with product_id:- {product_id}")
+                item.get('metadata')['id'] = product_id
+                LOG.info("Updating product metadata")
+                self.cgw_client.update_product(item.get('metadata'))
+            elif item.get('state') == 'absent' and product_id:
+                LOG.info(f"Deleting existing product records for product_id:- {product_id}")
+                self.cgw_client.delete_product(product_id)
+                LOG.info("Product record deleted!")
+            else:
+                LOG.warning("Invalid state or metadata!")
+        except Exception as err:
```

Comment:
We don't want to raise?

Reply:
I am not raising any error because I want other metadata to be operated. For example, if we have 10 metadata given in the YAML file and the state is invalid for the first metadata but all others are correct. In that case, it will raise an exception at the very first iteration and skip proceeding further. To iterate the further records I am logging a warning or it could have been LOG.error as well. I'll update the warning with an error. And I am open to discussing this design issue further, if you feel it must raise an exception I'm completely fine with your suggestion. Let me know what you think.

Reply:
I think for some pushes we do best effort, while for others we fail early. In any case, I think we need to fail at some point so that the user knows that their data is invalid. Users will probably not read our logs unless the push fails

Reply:
Will get this confirmed with release engineers and Jindrich as well for getting a better picture of this scenario.